File: backend/app/schemas/generation.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.correlation import RequirementCorrelationSummary
from app.schemas.quality import RequirementQualitySummary
from app.schemas.requirement import (
    RequirementGenerationMetadata,
    RequirementPriority,
    RequirementResponse,
    RequirementType,
)
from app.schemas.structured_requirement import StructuredRequirementFields


GenerationMode = Literal["feature", "decompose"]
RewriteGoal = Literal[
    "make_measurable",
    "improve_testability",
    "reduce_ambiguity",
    "clarify_units_conditions_scope",
    "decompose_wording",
]


class GeneratedRequirementCandidateBase(BaseModel):
    title: str = Field(min_length=3, max_length=255)
    text: str = Field(min_length=10)
    type: RequirementType
    priority: RequirementPriority
    rationale: str | None = None
    parent_requirement_id: str | None = Field(default=None, max_length=64)
    subsystem: str | None = Field(default=None, max_length=128)
    verification_method: str | None = Field(default=None, max_length=64)
    assumptions: str | None = None
    generation_metadata: RequirementGenerationMetadata


class GeneratedRequirementCandidateValidation(BaseModel):
    quality_summary: RequirementQualitySummary
    parsed_requirement: StructuredRequirementFields
    correlation_summary: RequirementCorrelationSummary


class GeneratedRequirementCandidateReview(GeneratedRequirementCandidateBase):
    temp_id: str = Field(min_length=1, max_length=64)
    suggested_id: str = Field(min_length=1, max_length=64)
    suggested_hierarchy: str | None = Field(default=None, max_length=128)
    validation: GeneratedRequirementCandidateValidation
    draft_label: str


class RequirementGenerationRequest(BaseModel):
    project_id: str = Field(min_length=1, max_length=64)
    mode: GenerationMode
    feature_description: str | None = Field(default=None, min_length=3)
    source_requirement_id: str | None = Field(default=None, max_length=64)


class GeneratedRequirementCandidateInput(GeneratedRequirementCandidateBase):
    temp_id: str = Field(min_length=1, max_length=64)
    suggested_id: str = Field(min_length=1, max_length=64)
    suggested_hierarchy: str | None = Field(default=None, max_length=128)


class RequirementGenerationReviewPayload(BaseModel):
    project_id: str = Field(min_length=1, max_length=64)
    candidates: list[GeneratedRequirementCandidateInput]


class RequirementGenerationSaveCandidate(GeneratedRequirementCandidateBase):
    temp_id: str = Field(min_length=1, max_length=64)


class RequirementGenerationSaveRequest(BaseModel):
    project_id: str = Field(min_length=1, max_length=64)
    candidates: list[RequirementGenerationSaveCandidate]


class RequirementGenerationResponse(BaseModel):
    project_id: str
    mode: GenerationMode
    source_requirement_id: str | None = None
    candidates: list[GeneratedRequirementCandidateReview]


class RequirementGenerationSaveResponse(BaseModel):
    saved_requirements: list[RequirementResponse]


class RequirementRewriteSuggestionItem(BaseModel):
    title: str
    text: str
    rationale: str
    generation_metadata: RequirementGenerationMetadata
    validation: GeneratedRequirementCandidateValidation
    draft_label: str


class RequirementRewriteSuggestionRequestPayload(BaseModel):
    project_id: str = Field(min_length=1, max_length=64)
    requirement_id: str | None = Field(default=None, max_length=64)
    title: str = Field(min_length=3, max_length=255)
    text: str = Field(min_length=10)
    type: RequirementType
    goals: list[RewriteGoal] = Field(default_factory=list)


class RequirementRewriteSuggestionResponse(BaseModel):
    provider: str | None = None
    suggestions: list[RequirementRewriteSuggestionItem]

File: frontend/src/types/generation.ts
import type {
  Requirement,
  RequirementCorrelationSummary,
  RequirementGenerationMetadata,
  RequirementPriority,
  RequirementQualitySummary,
  RequirementType,
  StructuredRequirementFields,
} from "./requirements";

export type GenerationMode = "feature" | "decompose";

export type GeneratedRequirementCandidateValidation = {
  quality_summary: RequirementQualitySummary;
  parsed_requirement: StructuredRequirementFields;
  correlation_summary: RequirementCorrelationSummary;
};

export type GeneratedRequirementCandidateReview = {
  temp_id: string;
  suggested_id: string;
  suggested_hierarchy: string | null;
  title: string;
  text: string;
  type: RequirementType;
  priority: RequirementPriority;
  rationale: string | null;
  parent_requirement_id: string | null;
  subsystem: string | null;
  verification_method: string | null;
  assumptions: string | null;
  generation_metadata: RequirementGenerationMetadata;
  validation: GeneratedRequirementCandidateValidation;
  draft_label: string;
};

export type RequirementGenerationResponse = {
  project_id: string;
  mode: GenerationMode;
  source_requirement_id: string | null;
  candidates: GeneratedRequirementCandidateReview[];
};

export type RequirementGenerationRequest = {
  project_id: string;
  mode: GenerationMode;
  feature_description?: string | null;
  source_requirement_id?: string | null;
};

export type GeneratedRequirementCandidateInput = {
  temp_id: string;
  suggested_id: string;
  suggested_hierarchy?: string | null;
  title: string;
  text: string;
  type: RequirementType;
  priority: RequirementPriority;
  rationale: string | null;
  parent_requirement_id: string | null;
  subsystem: string | null;
  verification_method: string | null;
  assumptions: string | null;
  generation_metadata: RequirementGenerationMetadata;
};

export type RequirementGenerationReviewRequest = {
  project_id: string;
  candidates: GeneratedRequirementCandidateInput[];
};

export type RequirementGenerationSaveCandidate = Omit<GeneratedRequirementCandidateInput, "suggested_id">;

export type RequirementGenerationSaveRequest = {
  project_id: string;
  candidates: RequirementGenerationSaveCandidate[];
};

export type RequirementGenerationSaveResponse = {
  saved_requirements: Requirement[];
};

export type GeneratedRequirementDraftState = GeneratedRequirementCandidateReview & {
  draft_id: string;
  decision: "accepted" | "rejected";
};

export type RequirementRewriteSuggestionRequest = {
  project_id: string;
  requirement_id?: string | null;
  title: string;
  text: string;
  type: RequirementType;
  goals?: RewriteGoal[];
};

export type RewriteGoal =
  | "make_measurable"
  | "improve_testability"
  | "reduce_ambiguity"
  | "clarify_units_conditions_scope"
  | "decompose_wording";

export type RequirementRewriteSuggestion = {
  title: string;
  text: string;
  rationale: string;
  generation_metadata: RequirementGenerationMetadata;
  validation: GeneratedRequirementCandidateValidation;
  draft_label: string;
};

export type RequirementRewriteSuggestionResponse = {
  provider: RequirementGenerationMetadata["generation_provider"] | null;
  suggestions: RequirementRewriteSuggestion[];
};
